crates/bazed-core/src/word_boundary.rs

```diff
@@ -0,0 +1,249 @@
+        Self {
+            iter: iter.into_iter(),
+            prev: None,
+            current_offset: 0,
+            reversing,
+        }
+    }
+}
+
+impl<I: Iterator<Item = char>> Iterator for WordBoundaries<I> {
+    type Item = (usize, WordBoundaryType);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            self.current_offset += 1;
+            let cur = self.iter.next()?;
+            let prev = self.prev;
+            self.prev = Some(cur);
+            if let Some(prev) = prev {
+                let boundary = if self.reversing {
+                    WordBoundaryType::between(cur, prev)
+                } else {
+                    WordBoundaryType::between(prev, cur)
+                };
+                if let Some(boundary) = boundary {
+                    return Some((self.current_offset - 1, boundary));
+                }
+            }
+        }
+    }
+}
+
+/// Returns an iterator that reverses over chunks of the rope.
+fn iter_rope_chunks_reverse<T: IntervalBounds>(rope: &Rope, range: T) -> ReverseChunkIter {
+    let Interval { start, end } = range.into_interval(rope.len());
+    ReverseChunkIter {
+        cursor: Cursor::new(rope, end),
+        cursor_at_end: true,
+        min: start,
+    }
+}
+
+/// More or less a copy of [xi_rope::rope::ChunkIter], but traversing the rope in reverse order
+struct ReverseChunkIter<'a> {
+    cursor: Cursor<'a, RopeInfo>,
+    cursor_at_end: bool,
+    min: usize,
+}
+
+impl<'a> Iterator for ReverseChunkIter<'a> {
+    type Item = &'a str;
+
+    fn next(&mut self) -> Option<&'a str> {
+        if self.cursor.pos() < self.min {
+            return None;
+        }
+        let (leaf, offset_within_leaf) = self.cursor.get_leaf()?;
+        let start_offset_of_leaf = self.cursor.pos() - offset_within_leaf;
+        let start_pos = self.min.saturating_sub(start_offset_of_leaf);
+        self.cursor.prev_leaf();
+        if self.cursor_at_end {
+            self.cursor_at_end = false;
+            Some(&leaf[start_pos..offset_within_leaf])
+        } else {
+            Some(&leaf[start_pos..])
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use xi_rope::Rope;
+
+    use super::WordBoundaries;
+    use crate::{
+        test_util,
+        word_boundary::{find_word_boundaries, find_word_boundaries_backwards, WordBoundaryType},
+    };
+
+    #[test]
+    fn test_word_boundaries() {
+        test_util::setup_test();
+        use WordBoundaryType::*;
+        fn boundaries(s: &str) -> Vec<(usize, WordBoundaryType)> {
+            WordBoundaries::from_iter(false, s.chars()).collect()
+        }
+        let actual = boundaries("foo foo...");
+        assert_eq!(vec![(3, End), (4, Start), (7, Both)], actual);
+        let actual = boundaries(" foo ");
+        assert_eq!(vec![(1, Start), (4, End)], actual);
+
+        // TODO we should have configurable word separators to allow `_` to not break words if the users wants that
```

Comment:
Optimally something like `go to end of token` even.

Reply:
Ye, that'd be a future wish, I think